[Diagnostics] Improve the diagnostic for invalid pointer conversion for an autoclosure result type.

hborla opened the pull request:
Add constraint fix `AllowAutoClosurePointerConversion` and corresponding diagnostic `AutoClosurePointerConversionFailure`. When we discover that we're trying to do an inout-to-pointer conversion for an autoclosure result in `matchTypes`, add the constraint fix, which tries to do the conversion as if the pointer type is a regular function argument.

hborla committed:
[Diagnostics] Improve the diagnostic for invalid pointer conversion for an autoclosure result type.

Add constraint fix `AllowAutoClosurePointerConversion` and corresponding diagnostic
`AutoClosurePointerConversionFailure`. When we discover that we're trying to do an
inout-to-pointer conversion in `matchTypes`, add the constraint fix, which tries to do the
conversion as if the pointer type is a regular function argument.

hborla commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -777,6 +777,28 @@ class AutoClosureForwardingFailure final : public FailureDiagnostic {
+/// var i = 0
+/// foo(&i) // Invalid conversion to UnsafePointer
+/// \endcode
+class AutoClosurePointerConversionFailure final : public FailureDiagnostic {
```
I think we can consider this a contextual failure so parent should be `ContextualFailure` here which also has storage for both types.

xedin commented on lib/Sema/CSFix.h:
```diff
@@ -621,6 +625,28 @@ class AutoClosureForwarding final : public ConstraintFix {
                                        ConstraintLocator *locator);
 };
 
+class AllowAutoClosurePointerConversion final : public ConstraintFix {
```
Same way as diagnostic is contextual, fix could also be contextual via inheriting it from `ContextualMismatch` which takes care of storage. @sl added a new constructor there which accepts `kind` as a parameter too.

xedin reviewed (approved):
Looks great, thank you! I have left a couple of small comments inline.

hborla committed:
[Diagnostics] Use `ContextualFailure` and `ContextualMistatch` as base classes
for `AutoClosurePointerConversionFailure` and `AllowAutoClosurePointerConversion`
respectively.

hborla commented:
@swift-ci please smoke test

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6966,6 +6971,13 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
+    if (recordFix(fix))
+      return SolutionKind::Error;
+    return matchTypes(type1, type2, matchKind, subflags,
+        locator.withPathElement(ConstraintLocator::FunctionArgument));
```
I'm not sure this is right – when matching types, `ConstraintLocator::FunctionArgument` is used to indicate that we're matching parameter types for a function conversion, which can add `SK_FunctionConversion` to the solution's score. This probably isn't a big deal for this case given we're going to be producing an error anyway, but perhaps we should be removing the `ConstraintLocator::AutoclosureResult` path element from the locator instead?

xedin replied:
I think that’s okay here, since the intent is to indicate that this is an argument, less trouble then removing last element and re-creating locator :)

hamishknight reviewed (commented):
Nice!

hborla commented:
@swift-ci please smoke test linux

hborla commented:
@swift-ci please smoke test Linux
